examples/logo_interpreter.rs

```diff
@@ -0,0 +1,87 @@
+//! To run a LOGO program, run:
+//!     cargo run --example logo_interpreter -- my_logo_program.txt
+
+extern crate turtle;
+
+use turtle::Turtle;
+use std::env;
+use std::fs::File;
+use std::io::{self, BufRead, BufReader};
+
+fn main() {
+    let mut turtle = Turtle::new();
+
+    let args: Vec<String> = env::args().collect();
+
+    if args.len() != 2 {
+        println!("Interactive LOGO Mode.\nType the commands in stdin.\nPress '^C' (cmd/ctrl+C) to exit");
+        let stdin = io::stdin();
+        let mut reader = BufReader::new(stdin);
+        loop {
+            let mut buffer = String::new();
+            let _ = reader
+                .read_line(&mut buffer)
+                .expect("Unable read input from stdin");
+                let cmd_args = buffer.split_whitespace();
+                handle_command(&mut turtle, cmd_args);
+        }
+    }
+
+    let path = &args[1];
+    let file = File::open(path).expect("Could not open provided program");
+    let mut reader = BufReader::new(file);
+    loop {
+        let mut buffer = String::new();
+        let read_bytes = reader
+            .read_line(&mut buffer)
+            .expect("Unable read input file");
+        if read_bytes == 0 {
+            // Reached EOF, stop interpreter
+            break;
+        }
+
+        let cmd_args = buffer.split_whitespace();
+        handle_command(&mut turtle, cmd_args);
+    }
```

Comment:
Great job with the repl @Jay9596! Using some of the generics and traits knowledge you now have from `handle_command`, let's remove the duplication from this code too: :smile: 

```rust
use turtle::Turtle;
use std::env;
use std::fs::File;
use std::io::{self, Read, BufRead, BufReader};

fn main() {
    let mut turtle = Turtle::new();

    let args: Vec<String> = env::args().collect();

    if args.len() != 2 {
        let stdin = io::stdin();
        interpret(&mut turtle, stdin);
    }
    else {
        let file = File::open(&args[1]).expect("Could not open provided program");
        interpret(&mut turtle, file);
    }

}

fn interpret<R: Read>(turtle: &mut Turtle, file: R) {
    let mut reader = BufReader::new(file);
    loop {
        let mut buffer = String::new();
        let read_bytes = reader.read_line(&mut buffer).expect("Unable read input");
        if read_bytes == 0 {
            // Reached EOF, stop interpreter
            break;
        }

        let cmd_args = buffer.split_whitespace();
        handle_command(turtle, cmd_args);
    }
}
```

Once again, let me know if you have any questions. You may want to take a look at how [`BufReader::new()`](https://doc.rust-lang.org/std/io/struct.BufReader.html) is defined.

Reply:
I was thinking of moving the read and loop body to a new `fn` like: `fn read_input<R: io::Read>(mut turtle: &mut Turtle, input: R)` . I have to get a copy of turtle as I can't return a `BufReader` and even if I do, the read_bytes needs to know the size at compile time, which can be different for `std::io::Stdin` and `std::fs::File`

Should I proceed with this, or is there a better way to do it ?

Reply:
I outlined a way to deal with all of that in my comment above. My code seems to address all of your concerns. Let me know if I'm misunderstanding what you're asking.

Reply:
Were you able to understand what I did in the code above?